crates/sats/src/uuid.rs

```diff
@@ -0,0 +1,381 @@
+    /// set the appropriate bits for the UUID version and variant.
+    ///
+    /// # Example
+    /// ```
+    /// # use spacetimedb_sats::uuid::Uuid;
+    /// // Use the `ReducerContext::rng()` method to generate random bytes in reducers,
+    /// // or call `ReducerContext::new_uuid_v4`
+    /// let random_bytes = [0u8; 16];
+    /// let uuid = Uuid::from_random_bytes_v4(random_bytes);
+    ///
+    /// assert_eq!(
+    ///     "00000000-0000-4000-8000-000000000000",
+    ///     uuid.to_string(),
+    /// );
+    /// ```
+    pub fn from_random_bytes_v4(counter_random_bytes: [u8; 16]) -> Self {
+        Self {
+            __uuid__: Builder::from_random_bytes(counter_random_bytes).into_uuid().as_u128(),
+        }
+    }
+
+    /// Generate a UUID `v7` using a monotonic counter from 0 to 2^31-1, a timestamp, and 4 random bytes.
+    ///
+    /// The counter will wrap around on overflow.
+    ///
+    /// The UUID `v7` is structured as follows:
+    ///```ascii
+    /// ┌───────────────────────────────────────────────┬───────────────────┐
+    /// | B0  | B1  | B2  | B3  | B4  | B5              |         B6        |
+    /// ├───────────────────────────────────────────────┼───────────────────┤
+    /// |                 unix_ts_ms                    |      version 7    |
+    /// └───────────────────────────────────────────────┴───────────────────┘
+    /// ┌──────────────┬─────────┬──────────────────┬───────────────────────┐
+    /// | B7           | B8      | B9  | B10 | B11  | B12 | B13 | B14 | B15 |
+    /// ├──────────────┼─────────┼──────────────────┼───────────────────────┤
+    /// | counter_high | variant |    counter_low   |        random         |
+    /// └──────────────┴─────────┴──────────────────┴───────────────────────┘
+    /// ```
+    /// # Panics
+    ///
+    /// Panics if the counter value is negative, or the timestamp is before the Unix epoch.
+    ///
+    /// # Example
+    ///```
+    /// use spacetimedb_sats::uuid::Uuid;
+    /// use spacetimedb_sats::timestamp::Timestamp;
+    ///
+    /// let now = Timestamp::from_micros_since_unix_epoch(1_686_000_000_000);
+    /// let counter = std::cell::RefCell::new(1);
+    /// // Use the `ReducerContext::rng()` | `ProcedureContext::rng()` to generate random bytes,
+    /// // or call `ReducerContext::new_uuid_v7()` / `ProcedureContext::new_uuid_v7()`
+    /// let random_bytes = [0u8; 4];
+    /// let uuid = Uuid::from_counter_v7(&counter, now, &random_bytes).unwrap();
+    ///
+    /// assert_eq!(
+    ///     "0000647e-5180-7000-8000-000200000000",
+    ///     uuid.to_string(),
+    /// );
+    /// ```
+    pub fn from_counter_v7(counter: &RefCell<i32>, now: Timestamp, random_bytes: &[u8; 4]) -> anyhow::Result<Self> {
+        // Monotonic counter value (31 bits)
+        let mut current = counter.borrow_mut();
+        // The counter is constructed from Timestamp::UNIX_EPOCH by callers
+        assert!(*current >= 0, "uuid counter must be non-negative");
```

Comment:
Why does this matter? And if the counter must be non-negative, why isn't it a `u32`?